### ISSUE TITLE
torch.stack not behaving properly on mps leads to Dataloader yielding same target values each batch

### ISSUE THREAD

**Issue by joulewei** [pytorch/pytorch#92265]
### 🐛 Describe the bug

When feeding a training loop with data using ```torch.utils.data.TensorDataset```and  ```torch.utils.data.DataLoader``` on an MPS device, the target data won't update during the iterations.

This is shown in the example below. Instead of ```1, 2, 3``` the values ```1, 1, 1``` appear as targets.

This happens with Pytorch **1.3.1** and **2.0.0.dev20230116**

```python
import torch
from torch.utils.data import TensorDataset, DataLoader

if torch.backends.mps.is_available():
    device = torch.device("mps")
    print("MPS is available")
else:
    device = torch.device("cpu")
    print("MPS is not available")

inputs = torch.randn(3, 1, device=device)
targets = torch.tensor([1, 2, 3], device=device)

dataset = TensorDataset(inputs, targets)

dataloader = DataLoader(dataset, batch_size=1)

# Iterating over the dataloader
for i, (input, target) in enumerate(dataloader):
    print(
        f'DataLoader: batch {i} => target {target.item()} but should be {i+1}')
```

MPS output:
```
MPS is available, running on mps
DataLoader: batch 0 => target 1 but should be 1
DataLoader: batch 1 => target 1 but should be 2
DataLoader: batch 2 => target 1 but should be 3
```

CPU output:
```
MPS is not available, running on cpu
DataLoader: batch 0 => target 1 but should be 1
DataLoader: batch 1 => target 2 but should be 2
DataLoader: batch 2 => target 3 but should be 3
```

### Versions

[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] numpy                     1.23.5           py39h1398885_0
[conda] numpy-base                1.23.5           py39h90707a3_0
[conda] pytorch                   1.13.1                  py3.9_0    pytorch
[conda] torchaudio                0.13.1                 py39_cpu    pytorch
[conda] torchvision               0.14.1                 py39_cpu    pytorch

cc @SsnL @VitalyFedyunin @ejguan @NivekT @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by ejguan** [comment]
I don't have mps device available now. Do you mind helping me test the following script?
```py
import torch

targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
for i in range(3):
    print(targets[i])
```

**Comment by joulewei** [comment]
Sure!

```python
>>> import torch
>>>
>>> targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
>>> for i in range(3):
...     print(targets[i])
...
tensor(1, device='mps:0')
tensor(2, device='mps:0')
tensor(3, device='mps:0')
```

I also iterated over the TensorDataset.

```python
>>> import torch
>>> from torch.utils.data import TensorDataset
>>>
>>>
>>> inputs = torch.randn(3, 1, device=torch.device("mps"))
>>> targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
>>>
>>> dataset = TensorDataset(inputs, targets)
>>>
>>> for (input, target) in dataset:
...     print(
...         f'DataLoader: target {target}')
...
DataLoader: target 1
DataLoader: target 2
DataLoader: target 3
```

**Comment by ejguan** [comment]
How about
```py
>>> targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
>>> for i in range(3):
...     print(targets[[i]])
```

**Comment by ejguan** [comment]
Thanks for providing the result. It has to be some logic within DataLoader unveil some problem for ops of mps tensor.
How about
```py
from torch.utils.data._utils.collate import default_collate
...
dataset = TensorDataset(inputs, targets)
for data in dataset:
    print(dafault_collate([data]))
```

**Comment by joulewei** [comment]
```python
>>> import torch
>>> from torch.utils.data import TensorDataset
>>> from torch.utils.data._utils.collate import default_collate
>>>
>>> inputs = torch.randn(3, 1, device=torch.device("mps"))
>>>targets = torch.tensor([1, 2, 3], device=torch.device("mps"))

>>>dataset = TensorDataset(inputs, targets)
>>>for data in dataset:
...    print(default_collate([data]))
>>> targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
>>>
>>> dataset = TensorDataset(inputs, targets)
>>> for data in dataset:
...     print(default_collate([data]))


/opt/homebrew/Caskroom/miniconda/base/envs/mlph/lib/python3.9/site-packages/torch/_tensor_str.py:115: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1670525849783/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(
[tensor([[-0.7102]], device='mps:0'), tensor([1], device='mps:0')]
[tensor([[-0.5577]], device='mps:0'), tensor([1], device='mps:0')]
[tensor([[0.0127]], device='mps:0'), tensor([1], device='mps:0')]
```

**Comment by ejguan** [comment]
Great, we have been able to narrow down the root cause.
How about:
```py
targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
for i in range(3):
    out = None
    print(torch.stack([targets[i]], 0, out=out))
    print(torch.stack([targets[i]], 0)
```

**Comment by joulewei** [comment]
Now I got lost :)

```python
>>> import torch
>>>
>>> targets = torch.tensor([1, 2, 3], device=torch.device("mps"))
>>> for i in range(3):
...     out = None
...     print(torch.stack([targets[i]], 0, out=out))
...     print(torch.stack([targets[i]], 0))
...
tensor([1], device='mps:0')
tensor([1], device='mps:0')
tensor([1], device='mps:0')
tensor([1], device='mps:0')
tensor([1], device='mps:0')
tensor([1], device='mps:0')
```

**Comment by joulewei** [comment]
I thinks it is the same as #87856

**Comment by DenisVieriu97** [comment]
Hi - this is fixed in latest pytorch nightly (torch 2.0.0.dev20230215). Please give it a try and let me know if you still any issues
```
torch 2.0.0.dev20230215
```
```
./python3 run.py
MPS is available
DataLoader: batch 0 => target 1 but should be 1
DataLoader: batch 1 => target 2 but should be 2
DataLoader: batch 2 => target 3 but should be 3
```

**Comment by joulewei** [comment]
Works!
